lib/tasker.js

```diff
@@ -0,0 +1,150 @@
+    this.#lastRoundUrl = roundUrl
+    res = await this.#fetch(`https://api.filspark.com${roundUrl}`, {
+      method: 'GET',
+      headers: { 'Content-Type': 'application/json' },
+      signal: AbortSignal.timeout(10_000)
+    })
+    await assertOkResponse(res, 'Failed to fetch the current SPARK round')
+    const { retrievalTasks, maxTasksPerNode, ...round } = await res.json()
+    console.log('Current SPARK round:', round)
+    console.log('  %s max tasks per round', maxTasksPerNode ?? '<n/a>')
+    console.log('  %s retrieval tasks', retrievalTasks.length)
+    this.maxTasksPerRound = maxTasksPerNode
+
+    const randomness = await getRandomnessForSparkRound(round.startEpoch)
+    console.log('  randomness: %s', randomness)
+
+    this.#remainingRoundTasks = await pickTasksForNode({
+      tasks: retrievalTasks,
+      maxTasksPerRound: this.maxTasksPerRound,
+      randomness,
+      stationId: Zinnia.stationId
+    })
+  }
+}
+
+const textEncoder = new TextEncoder()
+
+/**
+ * @param {Task} task
+ * @param {string} randomness
+ * @returns
+ */
+export async function getTaskKey (task, randomness) {
+  assertEquals(typeof task, 'object', 'task must be an object')
+  assertEquals(typeof task.cid, 'string', 'task.cid must be a string')
+  assertEquals(typeof task.minerId, 'string', 'task.minerId must be a string')
+  assertEquals(typeof randomness, 'string', 'randomness must be a string')
+
+  const data = [task.cid, task.minerId, randomness].join('\n')
+  const hash = await crypto.subtle.digest('sha-256', textEncoder.encode(data))
+  return BigInt('0x' + encodeHex(hash))
+}
+
+/**
+ * @param {string} stationId
+ */
+export async function getStationKey (stationId) {
+  assertEquals(typeof stationId, 'string', 'stationId must be a string')
+
+  const hash = await crypto.subtle.digest('sha-256', textEncoder.encode(stationId))
+  return BigInt('0x' + encodeHex(hash))
+}
+
+/**
+ * @param {object} args
+ * @param {Task[]} args.tasks
+ * @param {string} args.stationId
+ * @param {string} args.randomness
+ * @param {number} args.maxTasksPerRound
+ * @returns {Promise<Task[]>}
+ */
+export async function pickTasksForNode ({ tasks, stationId, randomness, maxTasksPerRound }) {
+  assertInstanceOf(tasks, Array, 'tasks must be an array')
+  assertEquals(typeof stationId, 'string', 'stationId must be a string')
+  assertEquals(typeof randomness, 'string', 'randomness must be a string')
+  assertEquals(typeof maxTasksPerRound, 'number', 'maxTasksPerRound must be a number')
+
+  const keyedTasks = await Promise.all(tasks.map(
+    async (t) => ({ ...t, key: await getTaskKey(t, randomness) })
+  ))
+  const stationKey = await getStationKey(stationId)
+
+  /**
+   * @param {{key: bigint}} a
+   * @param {{key: bigint}} b
+   * @returns {number}
+   */
+  const comparator = (a, b) => {
+    const ad = a.key ^ stationKey
+    const bd = b.key ^ stationKey
+    return ad > bd ? 1 : ad < bd ? -1 : 0
+  }
```

Comment:
instead of creating an inline function, what about

```js
const createComparator = stationKey => (a, b) => {
  // ...
}
```

Reply:
I prefer to keep the comparator as an inline function:
- It's short and succinct.
- When inside `pickTasksForNode`, we can keep the name short, because its clear what is the purpose of the function.
- `createComparator` is too generic, the name does not communicate what is being compared. We would need a name like `createComparatorForDistanceOfTaskToStation`, which is awfully long.